src/MergeContext.js

```diff
@@ -198,19 +217,269 @@ class StatusChecks
+class Labels
+{
+    // the labels parameter is the label array received from GitHub
+    constructor(labels, prNum) {
+        this._prNum = prNum;
+        this._labels = labels.map(label => new Label(label.name, true));
+    }
+
+    // adding a previously added or existing label is a no-op
+    add(name) {
+        const label = this._find(name);
+        if (label)
+            label.markForAddition();
+        else
+            this._labels.push(new Label(name, false));
+    }
+
+    // removing a previously removed or missing label is a no-op
+    remove(name) {
+        const label = this._find(name);
+        if (label)
+            label.markForRemoval();
+    }
+
+    // whether the label is present (from high-level Anubis code point of view)
+    has(name) {
+        const label = this._find(name);
+        return label && label.present();
+    }
+
+    // brings GitHub labels in sync with ours
+    async pushToGitHub() {
+        let syncedLabels = [];
+        for (let label of this._labels) {
+            if (label.needsRemovalFromGitHub()) {
+                await this._removeFromGitHub(label.name);
+            } else {
+                if (label.needsAdditionToGitHub())
+                    await this._addToGitHub(label.name); // TODO: Optimize to add all labels at once
+                // else still unchanged
+
+                syncedLabels.push(label);
+            }
+        }
+        this._labels = syncedLabels;
+    }
+
+    // a summary of changed labels (used for debugging)
+    diff() {
+        let str = "";
+        for (let label of this._labels) {
+            if (label.needsRemovalFromGitHub() || label.needsAdditionToGitHub()) {
+                if (str.length)
+                    str += ", ";
+                const prefix = label.present() ? '+' : '-';
+                str += prefix + label.name;
+            }
+        }
+        return '[' + str + ']';
+    }
+
+    // removes a single label from GitHub
+    async _removeFromGitHub(name) {
+        try {
+            await GH.removeLabel(name, this._prNum);
+        } catch (e) {
+            if (e.name === 'ErrorContext' && e.notFound()) {
+                Log.LogException(e, "_removeFromGitHub: " + name + " not found");
+                return;
+            }
+            throw e;
+        }
+    }
+
+    // adds a single label to GitHub
+    async _addToGitHub(name) {
+        let params = Util.commonParams();
+        params.number = this._prNum;
+        params.labels = [];
+        params.labels.push(name);
+
+        await GH.addLabels(params);
+    }
+
+    _find(name) { return this._labels.find(label => label.name === name); }
+}
+
+// A state of an open pull request (with regard to merging progress). One of:
+// brewing:
+//   a) without a staged commit (PRs are created in this state);
+//   b) with a fresh staged commit having failed checks, which is kept until it becomes stale
```

Comment:
Do not mix definitions with consequences/actions based on those definitions:

```suggestion
//   b) with a fresh staged commit having failed checks
```